eventdata/schedulers/utilization_scheduler.py

```diff
@@ -0,0 +1,79 @@
+# You may obtain a copy of the License at
+#
+#	http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import time
+import logging
+import random
+import statistics
+
+
+class UtilizationBasedScheduler:
+    """
+    This scheduler schedules events at 100% utilization (unthrottled) during the warmup time-period. It tracks this
+    period itself (i.e. independently of Rally) using the task parameter ``warmup-time-period``. During this period
+    it gathers response time metrics. The median response time and the provided target utilization (via the task
+    parameter ``target-utilization``) determine the average waiting time during the actual measurement phase of the
+    benchmark. In order to avoid that clients coordinate, we randomize waiting time using a Poisson distribution.
+    """
+    def __init__(self, params, perf_counter=time.perf_counter):
+        self.logger = logging.getLogger(__name__)
+        self.perf_counter = perf_counter
+        self.target_utilization = float(params["target-utilization"])
+        if self.target_utilization <= 0.0 or self.target_utilization > 1.0:
+            raise ValueError("target-utilization must be in the range (0.0, 1.0] but is {}".format(
+                self.target_utilization))
+        self.warmup_time_period = int(params["warmup-time-period"])
+        # to determine the target utilization
+        self.response_times = []
+        self.start_warmup = None
+        self.end_warmup = None
+        self.in_warmup = None
+        self.last_request_start = None
+        # determined by the utilization calculation
+        self.wait_time = None
+
+    def next(self, current):
+        if self.in_warmup is None:
+            self.in_warmup = True
+            self.start_warmup = self.perf_counter()
+            self.end_warmup = self.start_warmup + self.warmup_time_period
+            self.last_request_start = self.start_warmup
+            return 0
+        elif self.in_warmup:
+            now = self.perf_counter()
+            self.response_times.append((now - self.last_request_start))
+            self.last_request_start = now
+            if now >= self.end_warmup:
+                self.in_warmup = False
+                median_response_time_at_full_utilization = statistics.median(self.response_times)
+                # the total response time at the targeted utilization is determined by self.relative_target. However
+                # to determine the waiting time we need to subtract the (expected) response time from the total expected
+                # response time.
+                self.wait_time = median_response_time_at_full_utilization * (1 / self.target_utilization - 1)
```

Comment:
I think precedence of division over addition / subraction is a well-known concept from mathematics so I'm not sure we need to be that defensive here?

Reply:
I guess it's a matter of preference and I am fine either way. Personally I like being explicit because in maths it's inherently clearer due to the fraction bar as opposed to inline.

Reply:
Ok, you've convinced me. :) Let's get rid of some mental burden as this piece of code may be a bit hard to understand anyway.